### ISSUE TITLE
Flux.repeatWhen fails with Spec. Rule 1.3 - onSubscribe, onNext, onError and onComplete signaled to a Subscriber MUST be signaled serially.

### ISSUE THREAD

**Issue by kushagraThapar** [reactor/reactor-core#2579]
After updating to reactor-core 3.4.x, we are seeing `Flux.repeatWhen` fails with Spec. Rule 1.3 - onSubscribe, onNext, onError and onComplete signaled to a Subscriber MUST be signaled serially. 

## Expected Behavior
`Flux.repeatWhen` should not fail with the Spec. Rule 1.3, rather should work normally. 

## Actual Behavior
After updating to reactor-core 3.4.x, we are seeing `Flux.repeatWhen` fails with Spec. Rule 1.3 - onSubscribe, onNext, onError and onComplete signaled to a Subscriber MUST be signaled serially. 
This happens intermittently, because of some race condition, hard to reproduce, but every now and then, we see the below error: 

```
2021-01-26T18:26:28.4813099Z reactor.core.publisher.Sinks$EmissionException: Spec. Rule 1.3 - onSubscribe, onNext, onError and onComplete signaled to a Subscriber MUST be signaled serially.
2021-01-26T18:26:28.4814704Z 	at reactor.core.publisher.InternalManySink.emitNext(InternalManySink.java:56) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4815763Z 	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
2021-01-26T18:26:28.4816642Z Assembly trace from producer [reactor.core.publisher.FluxCallable] :
2021-01-26T18:26:28.4817488Z 	reactor.core@3.4.1/reactor.core.publisher.Mono.flux(Mono.java:2910)
2021-01-26T18:26:28.4819012Z 	com.azure.cosmos@4.12.0-beta.1/com.azure.cosmos.implementation.directconnectivity.StoreReader.readMultipleReplicasInternalAsync(StoreReader.java:356)
2021-01-26T18:26:28.4820078Z Error has been observed at the following site(s):
2021-01-26T18:26:28.4821145Z 	|_  â‡¢ reactor.core@3.4.1/reactor.core.publisher.Mono.flux(Mono.java:2910)
2021-01-26T18:26:28.4822052Z Stack trace:
2021-01-26T18:26:28.4823290Z 		at reactor.core.publisher.InternalManySink.emitNext(InternalManySink.java:56) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4872345Z 		at reactor.core.publisher.FluxRepeatWhen$RepeatWhenMainSubscriber.onComplete(FluxRepeatWhen.java:163) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4875181Z 		at reactor.core.publisher.FluxFlatMap$FlatMapMain.checkTerminated(FluxFlatMap.java:845) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4877042Z 		at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:607) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4878835Z 		at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:587) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4880171Z 		at reactor.core.publisher.FluxFlatMap$FlatMapMain.onComplete(FluxFlatMap.java:464) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4881495Z 		at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1785) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4882812Z 		at reactor.core.publisher.MonoSingle$SingleSubscriber.onComplete(MonoSingle.java:178) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4884198Z 		at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onComplete(FluxPeekFuseable.java:277) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4885562Z 		at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1785) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4886844Z 		at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:151) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4888146Z 		at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1784) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4889455Z 		at reactor.core.publisher.MonoSingle$SingleSubscriber.onComplete(MonoSingle.java:178) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4890927Z 		at reactor.core.publisher.FluxFlatMap$FlatMapMain.checkTerminated(FluxFlatMap.java:845) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4892683Z 		at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:607) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4894422Z 		at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:587) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4895780Z 		at reactor.core.publisher.FluxFlatMap$FlatMapMain.onComplete(FluxFlatMap.java:464) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4897227Z 		at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onComplete(FluxMapFuseable.java:150) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4898987Z 		at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1785) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4900453Z 		at reactor.core.publisher.FluxCallable.subscribe(FluxCallable.java:49) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4902239Z 		at reactor.core.publisher.InternalFluxOperator.subscribe(InternalFluxOperator.java:62) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4903587Z 		at reactor.core.publisher.FluxDefer.subscribe(FluxDefer.java:54) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4935008Z 		at reactor.core.publisher.InternalFluxOperator.subscribe(InternalFluxOperator.java:62) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4936644Z 		at reactor.core.publisher.FluxRepeatWhen$RepeatWhenMainSubscriber.resubscribe(FluxRepeatWhen.java:183) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4938350Z 		at reactor.core.publisher.FluxRepeatWhen$RepeatWhenOtherSubscriber.onNext(FluxRepeatWhen.java:238) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4939757Z 		at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:711) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4941085Z 		at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:587) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4942404Z 		at reactor.core.publisher.FluxFlatMap$FlatMapInner.onNext(FluxFlatMap.java:993) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4943745Z 		at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1766) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4945102Z 		at reactor.core.publisher.MonoDelayUntil$DelayUntilCoordinator.signal(MonoDelayUntil.java:261) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4946513Z 		at reactor.core.publisher.MonoDelayUntil$DelayUntilTrigger.onComplete(MonoDelayUntil.java:341) ~[reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4948078Z 		at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:120) [reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4949369Z 		at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68) [reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4950610Z 		at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28) [reactor-core-3.4.1.jar:?]
2021-01-26T18:26:28.4952029Z 		at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
2021-01-26T18:26:28.4956814Z 		at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
2021-01-26T18:26:28.4957954Z 		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
2021-01-26T18:26:28.4958915Z 		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
2021-01-26T18:26:28.4959767Z 		at java.lang.Thread.run(Thread.java:834) [?:?]
```

## Steps to Reproduce
I don't have a repro of this, as it is hard to get one. It is failing intermittently, but based on this previous issue: https://github.com/reactor/reactor-core/issues/2488 we can analyze the repro, and may be come up with one. 

## Possible Solution
Based on this earlier fix: https://github.com/reactor/reactor-core/pull/2499 , I think the same fix needs to be done for `Flux.repeatWhen` - as it is also facing the same race condition issue that was fixed in the above PR. 
Possible code change would be here : https://github.com/reactor/reactor-core/blob/master/reactor-core/src/main/java/reactor/core/publisher/FluxRepeatWhen.java#L162

## Your Environment
* Reactor version(s) used: 3.4.x
* JVM version (`java -version`): Java 8 and Java 11
* OS and version (eg `uname -a`): MacOS, Ubuntu 18.04

**Comment by kushagraThapar** [comment]
@simonbasle - any updates on this issue ?
We are blocked on this before moving to spring-boot 2.4.x because of this issue. Can you please provide a timeline for this fix ?

**Comment by bsideup** [comment]
@kushagraThapar thanks for reporting! The issue indeed is similar to `retryWhen`, I was able to reproduce and fix it in #2587 👍 

Will make it into the next release (see the milestones page for a schedule)

**Comment by kushagraThapar** [comment]
Thanks @bsideup for the quick fix, appreciate it.
